torch/distributed/_composable/fsdp/_fsdp_collectives.py

```diff
@@ -139,6 +139,10 @@ def foreach_reduce_scatter(
         # Record to mark the end of the reduce-scatter copy-in in the RS stream
         copy_in_event = torch.cuda.Event()
         copy_in_event.record()
+        # Only after the copy-in finishes can we free the gradients, which were
```

Comment:
FYI: I am sneaking this change into this PR for convenience. I ran the stream-related unit tests, and it does not affect correctness -- we do not need to be outside the stream context to have a stream wait for an event, so it does not matter that we move this up. The motivation is that we can clear the unsharded gradients _before_ we allocate the reduce-scatter output since we are only copying into the reduce-scatter input. This lowers peak memory by a small amount.